docs/Settings.md
# CLI Variable Reference

> Note: this document is autogenerated. Do not edit it manually.

### 3d_deadband_high

High value of throttle deadband for 3D mode (when stick is in the deadband range, the value in 3d_neutral is used instead)

| Default | Min | Max |
| --- | --- | --- |
| 1514 | PWM_RANGE_MIN | PWM_RANGE_MAX |

---

### 3d_deadband_low

Low value of throttle deadband for 3D mode (when stick is in the 3d_deadband_throttle range, the fixed values of 3d_deadband_low / _high are used instead)

| Default | Min | Max |
| --- | --- | --- |
| 1406 | PWM_RANGE_MIN | PWM_RANGE_MAX |

---

### 3d_deadband_throttle

Throttle signal will be held to a fixed value when throttle is centered with an error margin defined in this parameter.

| Default | Min | Max |
| --- | --- | --- |
| 50 | 0 | 200 |

---

### 3d_neutral

Neutral (stop) throttle value for 3D mode

| Default | Min | Max |
| --- | --- | --- |
| 1460 | PWM_RANGE_MIN | PWM_RANGE_MAX |

---

### acc_event_threshold_high

Acceleration threshold [cm/s/s] for impact / high g event text messages sent by SIM module. Acceleration values greater than 4 g can occur in fixed wing flight without an impact, so a setting of 4000 or greater is suggested. 0 = detection off.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 65535 |

---

### acc_event_threshold_low

Acceleration threshold [cm/s/s] for low-g / freefall detection text messages sent by SIM module. A setting of less than 100 is suggested. Valid values: [0-900], 0 = detection off.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 900 |

---

### acc_event_threshold_neg_x

Acceleration threshold [cm/s/s] for backwards acceleration / fixed wing landing detection text messages sent by SIM module. Suggested value for fixed wing: 1100. 0 = detection off.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 65535 |

---

### acc_hardware

Selection of acc hardware. See Wiki Sensor auto detect and hardware failure detection for more info

| Default | Min | Max |
| --- | --- | --- |
| AUTO |  |  |

---

### acc_lpf_hz

Software-based filter to remove mechanical vibrations from the accelerometer measurements. Value is cutoff frequency (Hz). For larger frames with bigger props set to lower value.

| Default | Min | Max |
| --- | --- | --- |
| 15 | 0 | 200 |

---

### acc_lpf_type

Specifies the type of the software LPF of the acc signals. BIQUAD gives better filtering and more delay, PT1 less filtering and less delay, so use only on clean builds.

| Default | Min | Max |
| --- | --- | --- |
| BIQUAD |  |  |

---

### acc_notch_cutoff

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 255 |

---

### acc_notch_hz

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 255 |

---

### accgain_x

Calculated value after '6 position avanced calibration'. Uncalibrated value is 4096. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 4096 | 1 | 8192 |

---

### accgain_y

Calculated value after '6 position avanced calibration'. Uncalibrated value is 4096. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 4096 | 1 | 8192 |

---

### accgain_z

Calculated value after '6 position avanced calibration'. Uncalibrated value is 4096. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 4096 | 1 | 8192 |

---

### acczero_x

Calculated value after '6 position avanced calibration'. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -32768 | 32767 |

---

### acczero_y

Calculated value after '6 position avanced calibration'. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -32768 | 32767 |

---

### acczero_z

Calculated value after '6 position avanced calibration'. See Wiki page.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -32768 | 32767 |

---

### ahrs_acc_ignore_rate

Total gyro rotation rate threshold [deg/s] before scaling to consider accelerometer trustworthy

| Default | Min | Max |
| --- | --- | --- |
| 15 | 0 | 30 |

---

### ahrs_acc_ignore_slope

Half-width of the interval to gradually reduce accelerometer weight. Centered at `imu_acc_ignore_rate` (exactly 50% weight)

| Default | Min | Max |
| --- | --- | --- |
| 5 | 0 | 10 |

---

### ahrs_dcm_ki

Inertial Measurement Unit KI Gain for accelerometer measurements

| Default | Min | Max |
| --- | --- | --- |
| 50 |  | 65535 |

---

### ahrs_dcm_ki_mag

Inertial Measurement Unit KI Gain for compass measurements

| Default | Min | Max |
| --- | --- | --- |
| 50 |  | 65535 |

---

### ahrs_dcm_kp

Inertial Measurement Unit KP Gain for accelerometer measurements

| Default | Min | Max |
| --- | --- | --- |
| 2000 |  | 65535 |

---

### ahrs_dcm_kp_mag

Inertial Measurement Unit KP Gain for compass measurements

| Default | Min | Max |
| --- | --- | --- |
| 2000 |  | 65535 |

---

### ahrs_gps_yaw_weight

Arhs gps yaw weight when mag is avaliable, 0 means no gps yaw, 100 means equal weight as compass

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 500 |

---

### ahrs_gps_yaw_windcomp

Wind compensation in heading estimation from gps groundcourse(fixed wing only)

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### ahrs_inertia_comp_method

Inertia force compensation method when gps is avaliable, VELNED use the accleration from gps, TURNRATE calculates accleration by turnrate multiplied by speed, ADAPTIVE choose best result from two in each ahrs loop

| Default | Min | Max |
| --- | --- | --- |
| ADAPTIVE |  |  |

---

### airmode_throttle_threshold

Defines airmode THROTTLE activation threshold when `airmode_type` **THROTTLE_THRESHOLD** is used

| Default | Min | Max |
| --- | --- | --- |
| 1150 | 1000 | 2000 |

---

### airmode_type

Defines the Airmode state handling type. Default **STICK_CENTER** is the classical approach in which Airmode is always active if enabled, but when the throttle is low and ROLL/PITCH/YAW sticks are centered, Iterms is not allowed to grow (ANTI_WINDUP). **THROTTLE_THRESHOLD** is the Airmode behavior known from Betaflight. In this mode, Airmode is active as soon THROTTLE position is above `airmode_throttle_threshold` and stays active until disarm. ANTI_WINDUP is never triggered. For small Multirotors (up to 7-inch propellers) it is suggested to switch to **THROTTLE_THRESHOLD** since it keeps full stabilization no matter what pilot does with the sticks. Fixed Wings always use **STICK_CENTER_ONCE** or **STICK_CENTER** modes.

| Default | Min | Max |
| --- | --- | --- |
| STICK_CENTER |  |  |

---

### airspeed_adc_channel

ADC channel to use for analog pitot tube (airspeed) sensor. If board doesn't have a dedicated connector for analog airspeed sensor will default to 0

| Default | Min | Max |
| --- | --- | --- |
| _target default_ | ADC_CHN_NONE | ADC_CHN_MAX |

---

### align_board_pitch

Arbitrary board rotation in deci-degrees (0.1 degree), to allow mounting it sideways / upside down / rotated etc

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_board_roll

Arbitrary board rotation in deci-degrees (0.1 degree), to allow mounting it sideways / upside down / rotated etc

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_board_yaw

Arbitrary board rotation in deci-degrees (0.1 degree), to allow mounting it sideways / upside down / rotated etc

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_mag

When running on non-default hardware or adding support for new sensors/sensor boards, these values are used for sensor orientation. When carefully understood, these values can also be used to rotate (in 90deg steps) or flip the board. Possible values are: DEFAULT, CW0_DEG, CW90_DEG, CW180_DEG, CW270_DEG, CW0_DEG_FLIP, CW90_DEG_FLIP, CW180_DEG_FLIP, CW270_DEG_FLIP.

| Default | Min | Max |
| --- | --- | --- |
| DEFAULT |  |  |

---

### align_mag_pitch

Same as align_mag_roll, but for the pitch axis.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_mag_roll

Set the external mag alignment on the roll axis (in 0.1 degree steps). If this value is non-zero, the compass is assumed to be externally mounted and both the board and on-board compass alignent (align_mag) are ignored. See also align_mag_pitch and align_mag_yaw.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_mag_yaw

Same as align_mag_roll, but for the yaw axis.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1800 | 3600 |

---

### align_opflow

Optical flow module alignment (default CW0_DEG_FLIP)

| Default | Min | Max |
| --- | --- | --- |
| CW0FLIP |  |  |

---

### alt_hold_deadband

Defines the deadband of throttle during alt_hold [r/c points]

| Default | Min | Max |
| --- | --- | --- |
| 50 | 10 | 250 |

---

### antigravity_accelerator

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 20 |

---

### antigravity_cutoff_lpf_hz

Antigravity cutoff frequenct for Throtte filter. Antigravity is based on the difference between actual and filtered throttle input. The bigger is the difference, the bigger Antigravity gain

| Default | Min | Max |
| --- | --- | --- |
| 15 | 1 | 30 |

---

### antigravity_gain

Max Antigravity gain. `1` means Antigravity is disabled, `2` means Iterm is allowed to double during rapid throttle movements

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 20 |

---

### applied_defaults

Internal (configurator) hint. Should not be changed manually

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 99 |

---

### baro_cal_tolerance

Baro calibration tolerance in cm. The default should allow the noisiest baro to complete calibration [cm].

| Default | Min | Max |
| --- | --- | --- |
| 150 | 0 | 1000 |

---

### baro_hardware

Selection of baro hardware. See Wiki Sensor auto detect and hardware failure detection for more info

| Default | Min | Max |
| --- | --- | --- |
| AUTO |  |  |

---

### bat_cells

Number of cells of the battery (0 = auto-detect), see battery documentation. 7S, 9S and 11S batteries cannot be auto-detected.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 12 |

---

### bat_voltage_src

Chose between raw and sag compensated battery voltage to use for battery alarms and telemetry. Possible values are `RAW` and `SAG_COMP`

| Default | Min | Max |
| --- | --- | --- |
| RAW |  |  |

---

### battery_capacity

Set the battery capacity in mAh or mWh (see `battery_capacity_unit`). Used to calculate the remaining battery capacity.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 4294967295 |

---

### battery_capacity_critical

If the remaining battery capacity goes below this threshold the battery is considered empty and the beeper will emit long beeps.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 4294967295 |

---

### battery_capacity_unit

Unit used for `battery_capacity`, `battery_capacity_warning` and `battery_capacity_critical` [MAH/MWH] (milliAmpere hour / milliWatt hour).

| Default | Min | Max |
| --- | --- | --- |
| MAH |  |  |

---

### battery_capacity_warning

If the remaining battery capacity goes below this threshold the beeper will emit short beeps and the relevant OSD items will blink.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 4294967295 |

---

### beeper_pwm_mode

Allows disabling PWM mode for beeper on some targets. Switch from ON to OFF if the external beeper sound is weak. Do not switch from OFF to ON without checking if the board supports PWM beeper mode

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### blackbox_device

Selection of where to write blackbox data

| Default | Min | Max |
| --- | --- | --- |
| _target default_ |  |  |

---

### blackbox_rate_denom

Blackbox logging rate denominator. See blackbox_rate_num.

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 65535 |

---

### blackbox_rate_num

Blackbox logging rate numerator. Use num/denom settings to decide if a frame should be logged, allowing control of the portion of logged loop iterations

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 65535 |

---

### control_deadband

Stick deadband in [r/c points], applied after r/c deadband and expo. Used to check if sticks are centered.

| Default | Min | Max |
| --- | --- | --- |
| 10 | 2 | 250 |

---

### controlrate_profile

Control rate profile to switch to when the battery profile is selected, 0 to disable and keep the currently selected control rate profile

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 3 |

---

### cpu_underclock

This option is only available on certain architectures (F3 CPUs at the moment). It makes CPU clock lower to reduce interference to long-range RC systems working at 433MHz

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### cruise_power

Power draw at cruise throttle used for remaining flight time/distance estimation in 0.01W unit

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 4294967295 |

---

### current_adc_channel

ADC channel to use for analog current sensor input. Defaults to board CURRENT sensor input (if available). 0 = disabled

| Default | Min | Max |
| --- | --- | --- |
| _target default_ | ADC_CHN_NONE | ADC_CHN_MAX |

---

### current_meter_offset

This sets the output offset voltage of the current sensor in millivolts.

| Default | Min | Max |
| --- | --- | --- |
| _target default_ | -32768 | 32767 |

---

### current_meter_scale

This sets the output voltage to current scaling for the current sensor in 0.1 mV/A steps. 400 is 40mV/A such as the ACS756 sensor outputs. 183 is the setting for the uberdistro with a 0.25mOhm shunt.

| Default | Min | Max |
| --- | --- | --- |
| _target default_ | -10000 | 10000 |

---

### current_meter_type

ADC , VIRTUAL, NONE. The virtual current sensor, once calibrated, estimates the current value from throttle position.

| Default | Min | Max |
| --- | --- | --- |
| ADC |  |  |

---

### d_boost_gyro_delta_lpf_hz

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 80 | 10 | 250 |

---

### d_boost_max

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 1.25 | 1 | 3 |

---

### d_boost_max_at_acceleration

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 7500 | 1000 | 16000 |

---

### d_boost_min

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 0.5 | 0 | 1 |

---

### deadband

These are values (in us) by how much RC input can be different before it's considered valid. For transmitters with jitter on outputs, this value can be increased. Defaults are zero, but can be increased up to 10 or so if rc inputs twitch while idle.

| Default | Min | Max |
| --- | --- | --- |
| 5 | 0 | 32 |

---

### debug_mode

Defines debug values exposed in debug variables (developer / debugging setting)

| Default | Min | Max |
| --- | --- | --- |
| NONE |  |  |

---

### disarm_kill_switch

Disarms the motors independently of throttle value. Setting to OFF reverts to the old behaviour of disarming only when the throttle is low. Only applies when arming and disarming with an AUX channel.

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### display_force_sw_blink

OFF = OSD hardware blink / ON = OSD software blink. If OSD warning text/values are invisible, try setting this to ON

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### dji_cn_alternating_duration

Alternating duration of craft name elements, in tenths of a second

| Default | Min | Max |
| --- | --- | --- |
| 30 | 1 | 150 |

---

### dji_esc_temp_source

Re-purpose the ESC temperature field for IMU/BARO temperature

| Default | Min | Max |
| --- | --- | --- |
| ESC |  |  |

---

### dji_message_speed_source

Sets the speed type displayed by the DJI OSD in craft name: GROUND, 3D, AIR

| Default | Min | Max |
| --- | --- | --- |
| 3D |  |  |

---

### dji_rssi_source

Source of the DJI RSSI field: RSSI, CRSF_LQ

| Default | Min | Max |
| --- | --- | --- |
| RSSI |  |  |

---

### dji_use_adjustments

Show inflight adjustments in craft name field

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### dji_use_name_for_messages

Re-purpose the craft name field for messages.

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### dji_workarounds

Enables workarounds for different versions of MSP protocol used

| Default | Min | Max |
| --- | --- | --- |
| 1 | 0 | 255 |

---

### dshot_beeper_enabled

Whether using DShot motors as beepers is enabled

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### dshot_beeper_tone

Sets the DShot beeper tone

| Default | Min | Max |
| --- | --- | --- |
| 1 | 1 | 5 |

---

### dterm_lpf_hz

Dterm low pass filter cutoff frequency. Default setting is very conservative and small multirotors should use higher value between 80 and 100Hz. 80 seems like a gold spot for 7-inch builds while 100 should work best with 5-inch machines. If motors are getting too hot, lower the value

| Default | Min | Max |
| --- | --- | --- |
| 110 | 0 | 500 |

---

### dterm_lpf_type

Defines the type of stage 1 D-term LPF filter. Possible values: `PT1`, `BIQUAD`, `PT2`, `PT3`.

| Default | Min | Max |
| --- | --- | --- |
| PT2 |  |  |

---

### dynamic_gyro_notch_3d_q

Q factor for 3D dynamic notches

| Default | Min | Max |
| --- | --- | --- |
| 200 | 1 | 1000 |

---

### dynamic_gyro_notch_enabled

Enable/disable dynamic gyro notch also known as Matrix Filter

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### dynamic_gyro_notch_min_hz

Minimum frequency for dynamic notches. Default value of `150` works best with 5" multirotors. Should be lowered with increased size of propellers. Values around `100` work fine on 7" drones. 10" can go down to `60` - `70`

| Default | Min | Max |
| --- | --- | --- |
| 50 | 30 | 250 |

---

### dynamic_gyro_notch_mode

Gyro dynamic notch type

| Default | Min | Max |
| --- | --- | --- |
| 2D |  |  |

---

### dynamic_gyro_notch_q

Q factor for dynamic notches

| Default | Min | Max |
| --- | --- | --- |
| 120 | 1 | 1000 |

---

### esc_sensor_listen_only

Enable when BLHeli32 Auto Telemetry function is used. Disable in every other case

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### ez_aggressiveness

EzTune aggressiveness

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### ez_axis_ratio

EzTune axis ratio

| Default | Min | Max |
| --- | --- | --- |
| 110 | 25 | 175 |

---

### ez_damping

EzTune damping

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### ez_enabled

Enables EzTune feature

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### ez_expo

EzTune expo

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### ez_filter_hz

EzTune filter cutoff frequency

| Default | Min | Max |
| --- | --- | --- |
| 110 | 10 | 300 |

---

### ez_rate

EzTune rate

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### ez_response

EzTune response

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### ez_stability

EzTune stability

| Default | Min | Max |
| --- | --- | --- |
| 100 | 0 | 200 |

---

### failsafe_delay

Time in deciseconds to wait before activating failsafe when signal is lost. See [Failsafe documentation](Failsafe.md#failsafe_delay).

| Default | Min | Max |
| --- | --- | --- |
| 5 | 0 | 200 |

---

### failsafe_fw_pitch_angle

Amount of dive/climb when `LAND` (or old `SET-THR`) failsafe is active on a fixed-wing machine. In 1/10 deg (deci-degrees). Negative values = climb

| Default | Min | Max |
| --- | --- | --- |
| 100 | -800 | 800 |

---

### failsafe_fw_roll_angle

Amount of banking when `LAND` (or old `SET-THR`) failsafe is active on a fixed-wing machine. In 1/10 deg (deci-degrees). Negative values = left roll

| Default | Min | Max |
| --- | --- | --- |
| -200 | -800 | 800 |

---

### failsafe_fw_yaw_rate

Requested yaw rate to execute when `LAND` (or old `SET-THR`) failsafe is active on a fixed-wing machine. In deg/s. Negative values = left turn

| Default | Min | Max |
| --- | --- | --- |
| -45 | -1000 | 1000 |

---

### failsafe_lights

Enable or disable the lights when the `FAILSAFE` flight mode is enabled. The target needs to be compiled with `USE_LIGHTS` [ON/OFF].

| Default | Min | Max |
| --- | --- | --- |
| ON | OFF | ON |

---

### failsafe_lights_flash_on_time

Flash lights ON time in milliseconds when `failsafe_lights` is ON and `FAILSAFE` flight mode is enabled. [20-65535].

| Default | Min | Max |
| --- | --- | --- |
| 100 | 20 | 65535 |

---

### failsafe_lights_flash_period

Time in milliseconds between two flashes when `failsafe_lights` is ON and `FAILSAFE` flight mode is enabled [40-65535].

| Default | Min | Max |
| --- | --- | --- |
| 1000 | 40 | 65535 |

---

### failsafe_min_distance

If failsafe happens when craft is closer than this distance in centimeters from home, failsafe will not execute regular failsafe_procedure, but will execute procedure specified in failsafe_min_distance_procedure instead. 0 = Normal failsafe_procedure always taken.

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 65000 |

---

### failsafe_min_distance_procedure

What failsafe procedure to initiate in Stage 2 when craft is closer to home than failsafe_min_distance. See [Failsafe documentation](Failsafe.md#failsafe_throttle).

| Default | Min | Max |
| --- | --- | --- |
| DROP |  |  |

---

### failsafe_mission_delay

Applies if failsafe occurs when a WP mission is in progress. Sets the time delay in seconds between failsafe occurring and the selected failsafe procedure activating. If set to -1 the failsafe procedure won't activate at all and the mission will continue until the end.

| Default | Min | Max |
| --- | --- | --- |
| 0 | -1 | 600 |

---

### failsafe_off_delay

Time in deciseconds to wait before turning off motors when failsafe is activated. 0 = No timeout. See [Failsafe documentation](Failsafe.md#failsafe_off_delay).

| Default | Min | Max |
| --- | --- | --- |
| 200 | 0 | 200 |

---

### failsafe_procedure

What failsafe procedure to initiate in Stage 2. See [Failsafe documentation](Failsafe.md#failsafe_throttle).

| Default | Min | Max |
| --- | --- | --- |
| LAND |  |  |

---

### failsafe_recovery_delay

Time in deciseconds to wait before aborting failsafe when signal is recovered. See [Failsafe documentation](Failsafe.md#failsafe_recovery_delay).

| Default | Min | Max |
| --- | --- | --- |
| 5 | 0 | 200 |

---

### failsafe_stick_threshold

Threshold for stick motion to consider failsafe condition resolved. If non-zero failsafe won't clear even if RC link is restored - you have to move sticks to exit failsafe.

| Default | Min | Max |
| --- | --- | --- |
| 50 | 0 | 500 |

---

### failsafe_throttle

Throttle level used for landing when failsafe is enabled. See [Failsafe documentation](Failsafe.md#failsafe_throttle).

| Default | Min | Max |
| --- | --- | --- |
| 1000 | PWM_RANGE_MIN | PWM_RANGE_MAX |

---

### failsafe_throttle_low_delay

If failsafe activated when throttle is low for this much time - bypass failsafe and disarm, in 10th of seconds. 0 = No timeout

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 300 |

---

### fixed_wing_auto_arm

Auto-arm fixed wing aircraft on throttle above min_check, and disarming with stick commands are disabled, so power cycle is required to disarm. Requires enabled motorstop and no arm switch configured.

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### flaperon_throw_offset

Defines throw range in us for both ailerons that will be passed to servo mixer via input source 14 (`FEATURE FLAPS`) when FLAPERON mode is activated.

| Default | Min | Max |
| --- | --- | --- |
| 200 | FLAPERON_THROW_MIN | FLAPERON_THROW_MAX |

---

### fpv_mix_degrees

_// TODO_

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 50 |

---

### frsky_pitch_roll

S.Port telemetry: Send pitch and roll degrees*10 instead of raw accelerometer data

| Default | Min | Max |
| --- | --- | --- |
| OFF | OFF | ON |

---

### fw_autotune_max_rate_deflection

The target percentage of maximum mixer output used for determining the rates in `AUTO` and `LIMIT`.

| Default | Min | Max |
| --- | --- | --- |
| 80 | 50 | 100 |

---

### fw_autotune_min_stick

Minimum stick input [%], after applying deadband and expo, to start recording the plane's response to stick input.

| Default | Min | Max |
| --- | --- | --- |
| 50 | 0 | 100 |

---

### fw_autotune_rate_adjustment

`AUTO` and `LIMIT` adjust the rates to match the capabilities of the airplane, with `LIMIT` they are never increased above the starting rates setting. `FIXED` does not adjust the rates. Rates are not changed when tuning in `ANGLE` mode.

| Default | Min | Max |
| --- | --- | --- |
| AUTO |  |  |

---

### fw_d_level

Fixed-wing attitude stabilisation HORIZON transition point

| Default | Min | Max |
| --- | --- | --- |
| 75 | 0 | 255 |

---

### fw_d_pitch

Fixed wing rate stabilisation D-gain for PITCH

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 255 |

---

### fw_d_roll

Fixed wing rate stabilisation D-gain for ROLL

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 255 |

---

### fw_d_yaw

Fixed wing rate stabilisation D-gain for YAW

| Default | Min | Max |
| --- | --- | --- |
| 0 | 0 | 255 |

---

### fw_ff_pitch

Fixed-wing rate stabilisation FF-gain for PITCH

| Default | Min | Max |
| --- | --- | --- |
| 50 | 0 | 255 |

---

### fw_ff_roll

Fixed-wing rate stabilisation FF-gain for ROLL

| Default | Min | Max |
| --- | --- | --- |
| 50 | 0 | 255 |

---

### fw_ff_yaw

Fixed-wing rate stabilisation FF-gain for YAW

| Default | Min | Max |
| --- | --- | --- |
| 60 | 0 | 255 |

---

### fw_i_level

Fixed-wing attitude stabilisation low-pass filter cutoff

| Default | Min | Max |
| --- | --- | --- |
| 5 | 0 | 255 |

---

### fw_i_pitch

Fixed-wing rate stabilisation I-gain for PITCH

| Default | Min | Max |
| --- | --- | --- |
| 7 | 0 | 255 |

---

### fw_i_roll

Fixed-wing rate stabilisation I-gain for ROLL

| Default | Min | Max |
| --- | --- | --- |
| 7 | 0 | 255 |

---

### fw_i_yaw

Fixed-wing rate stabilisation I-gain for YAW

| Default | Min | Max |
| --- | --- | --- |
| 10 | 0 | 255 |

---

### fw_iterm_limit_stick_position


src/main/fc/config.h
/*
 * This file is part of Cleanflight.
 *
 * Cleanflight is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Cleanflight is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
 */

#pragma once

#include <stdint.h>
#include <stdbool.h>
#include "common/time.h"
#include "config/parameter_group.h"
#include "drivers/adc.h"
#include "drivers/rx_pwm.h"
#include "fc/stats.h"

#define MAX_PROFILE_COUNT 3
#define ONESHOT_FEATURE_CHANGED_DELAY_ON_BOOT_MS 1500
#define MAX_NAME_LENGTH 16

#define ACC_TASK_FREQUENCY_DEFAULT 500
#define ACC_TASK_FREQUENCY_MIN 100
#define ACC_TASK_FREQUENCY_MAX 1000
#define ATTITUDE_TASK_FREQUENCY_DEFAULT 250
#define ATTITUDE_TASK_FREQUENCY_MIN 100
#define ATTITUDE_TASK_FREQUENCY_MAX 1000

typedef enum {
    ASYNC_MODE_NONE,
    ASYNC_MODE_GYRO,
    ASYNC_MODE_ALL
} asyncMode_e;

typedef enum {
    FEATURE_UNUSED_5 = 1 << 0,          // RX_PPM
    FEATURE_VBAT = 1 << 1,
    FEATURE_UNUSED_1 = 1 << 2,          // Unused in INAV
    FEATURE_UNUSED_6 = 1 << 3,          // RX_SERIAL
    FEATURE_MOTOR_STOP = 1 << 4,
    FEATURE_SERVO_TILT = 1 << 5,
    FEATURE_SOFTSERIAL = 1 << 6,
    FEATURE_GPS = 1 << 7,
    FEATURE_UNUSED_3 = 1 << 8,          // was FEATURE_FAILSAFE
    FEATURE_UNUSED_4 = 1 << 9,          // was FEATURE_SONAR
    FEATURE_TELEMETRY = 1 << 10,
    FEATURE_CURRENT_METER = 1 << 11,
    FEATURE_3D = 1 << 12,
    FEATURE_UNUSED_7 = 1 << 13,         // RX_PARALLEL_PWM
    FEATURE_UNUSED_8 = 1 << 14,         // RX_MSP
    FEATURE_RSSI_ADC = 1 << 15,
    FEATURE_LED_STRIP = 1 << 16,
    FEATURE_DASHBOARD = 1 << 17,
    FEATURE_UNUSED_2 = 1 << 18,         // Unused in INAV
    FEATURE_BLACKBOX = 1 << 19,
    FEATURE_CHANNEL_FORWARDING = 1 << 20,
    FEATURE_TRANSPONDER = 1 << 21,
    FEATURE_AIRMODE = 1 << 22,
    FEATURE_SUPEREXPO_RATES = 1 << 23,
    FEATURE_VTX = 1 << 24,
    FEATURE_UNUSED_9 = 1 << 25,         // RX_SPI
    FEATURE_UNUSED_10 = 1 << 26,        //SOFTSPI
    FEATURE_PWM_SERVO_DRIVER = 1 << 27,
    FEATURE_PWM_OUTPUT_ENABLE = 1 << 28,
    FEATURE_OSD = 1 << 29,
    FEATURE_DEBUG_TRACE = 1 << 30,
} features_e;

typedef struct systemConfig_s {
    uint16_t accTaskFrequency;
    uint16_t attitudeTaskFrequency;
    uint8_t current_profile_index;
    uint8_t asyncMode;
    uint8_t debug_mode;
    uint8_t i2c_speed;
    uint8_t cpuUnderclock;
    uint8_t throttle_tilt_compensation_strength;      // the correction that will be applied at throttle_correction_angle.
    inputFilteringMode_e pwmRxInputFilteringMode;
    char name[MAX_NAME_LENGTH + 1];
} systemConfig_t;

PG_DECLARE(systemConfig_t, systemConfig);

typedef struct beeperConfig_s {
    uint32_t beeper_off_flags;
    uint32_t preferred_beeper_off_flags;
} beeperConfig_t;

PG_DECLARE(beeperConfig_t, beeperConfig);

typedef struct adcChannelConfig_s {
    uint8_t adcFunctionChannel[ADC_FUNCTION_COUNT];
} adcChannelConfig_t;

PG_DECLARE(adcChannelConfig_t, adcChannelConfig);


#ifdef USE_STATS
PG_DECLARE(statsConfig_t, statsConfig);
#endif

void beeperOffSet(uint32_t mask);
void beeperOffSetAll(uint8_t beeperCount);
void beeperOffClear(uint32_t mask);
void beeperOffClearAll(void);
uint32_t getBeeperOffMask(void);
void setBeeperOffMask(uint32_t mask);
uint32_t getPreferredBeeperOffMask(void);
void setPreferredBeeperOffMask(uint32_t mask);

void copyCurrentProfileToProfileSlot(uint8_t profileSlotIndex);

void initEEPROM(void);
void resetEEPROM(void);
void readEEPROM(void);
void writeEEPROM(void);
void ensureEEPROMContainsValidData(void);

void saveConfigAndNotify(void);
void validateAndFixConfig(void);

uint8_t getConfigProfile(void);
bool setConfigProfile(uint8_t profileIndex);
void setConfigProfileAndWriteEEPROM(uint8_t profileIndex);

bool canSoftwareSerialBeUsed(void);
void applyAndSaveBoardAlignmentDelta(int16_t roll, int16_t pitch);

void createDefaultConfig(void);
void resetConfigs(void);
void targetConfiguration(void);

uint32_t getPidUpdateRate(void);
timeDelta_t getGyroUpdateRate(void);
uint16_t getAccUpdateRate(void);
#ifdef USE_ASYNC_GYRO_PROCESSING
uint16_t getAttitudeUpdateRate(void);
uint8_t getAsyncMode(void);
#endif
